[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#92035 [FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`**
* #91767 [FSDP] Do not clean FQNs even for `use_orig_params=True`
* #92031 [FSDP][BE] Improve `device_id` + CPU offload test
* #92028 [FSDP][BE] Rename `prefixed_param_names` -> `fqns` for consolidation
* #92027 [FSDP][BE] Better error msg for incorrect device for training

IIUC, FSDP assumes that rank `r` uses `cuda:r`. In other words, FSDP assumes single-process single-device (SPSD) and CUDA devices only.

The SPSD assumption can help us simplify the `compute_device` and `device_id` code since then specifying `device_id` is simply a signal to FSDP whether the user wants FSDP to move parameters on behalf of the user or not. The actual target device is fixed (namely, `cuda:r`). Since the SPSD assumption is deeply embedded into FSDP, I think this is reasonable. We can generalize over non-CUDA devices in the future while still enforcing SPSD.

Feel free to let me know your guys' thoughts.

If we go with this, we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device. See https://github.com/pytorch/pytorch/issues/91661.

Commit:
[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`

[ghstack-poisoned]

Commit:
Update on "[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`"


IIUC, FSDP assumes that rank `r` uses `cuda:r`. In other words, FSDP assumes single-process single-device (SPSD) and CUDA devices only.

The SPSD assumption can help us simplify the `compute_device` and `device_id` code since then specifying `device_id` is simply a signal to FSDP whether the user wants FSDP to move parameters on behalf of the user or not. The actual target device is fixed (namely, `cuda:r`). Since the SPSD assumption is deeply embedded into FSDP, I think this is reasonable. We can generalize over non-CUDA devices in the future while still enforcing SPSD.

Feel free to let me know your guys' thoughts.

If we go with this, we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device. See https://github.com/pytorch/pytorch/issues/91661.

[ghstack-poisoned]

zhaojuanmao commented:
I think the SPSD and CUDA device only assumption is current FSDP state, " we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device." sounds great to me, thank you!!

Commit:
Update on "[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`"


IIUC, FSDP assumes that rank `r` uses `cuda:r`. In other words, FSDP assumes single-process single-device (SPSD) and CUDA devices only.

The SPSD assumption can help us simplify the `compute_device` and `device_id` code since then specifying `device_id` is simply a signal to FSDP whether the user wants FSDP to move parameters on behalf of the user or not. The actual target device is fixed (namely, `cuda:r`). Since the SPSD assumption is deeply embedded into FSDP, I think this is reasonable. We can generalize over non-CUDA devices in the future while still enforcing SPSD.

Feel free to let me know your guys' thoughts.

If we go with this, we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device. See https://github.com/pytorch/pytorch/issues/91661.

[ghstack-poisoned]

Commit:
Update on "[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`"


IIUC, FSDP assumes that rank `r` uses `cuda:r`. In other words, FSDP assumes single-process single-device (SPSD) and CUDA devices only.

The SPSD assumption can help us simplify the `compute_device` and `device_id` code since then specifying `device_id` is simply a signal to FSDP whether the user wants FSDP to move parameters on behalf of the user or not. The actual target device is fixed (namely, `cuda:r`). Since the SPSD assumption is deeply embedded into FSDP, I think this is reasonable. We can generalize over non-CUDA devices in the future while still enforcing SPSD.

Feel free to let me know your guys' thoughts.

If we go with this, we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device. See https://github.com/pytorch/pytorch/issues/91661.

[ghstack-poisoned]

Commit:
Update on "[FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`"


IIUC, FSDP assumes that rank `r` uses `cuda:r`. In other words, FSDP assumes single-process single-device (SPSD) and CUDA devices only.

The SPSD assumption can help us simplify the `compute_device` and `device_id` code since then specifying `device_id` is simply a signal to FSDP whether the user wants FSDP to move parameters on behalf of the user or not. The actual target device is fixed (namely, `cuda:r`). Since the SPSD assumption is deeply embedded into FSDP, I think this is reasonable. We can generalize over non-CUDA devices in the future while still enforcing SPSD.

Feel free to let me know your guys' thoughts.

If we go with this, we can provide earlier and cleaner error handling in the case the user forgets to set the current CUDA device. See https://github.com/pytorch/pytorch/issues/91661.

[ghstack-poisoned]
